consensus/consensus-types/src/block_retrieval.rs

```diff
@@ -75,20 +100,25 @@ impl BlockRetrievalResponse {
 
     pub fn verify(
         &self,
-        block_id: HashValue,
-        num_blocks: u64,
+        retrieval_request: BlockRetrievalRequest,
         sig_verifier: &ValidatorVerifier,
     ) -> anyhow::Result<()> {
         ensure!(
             self.status != BlockRetrievalStatus::Succeeded
-                || self.blocks.len() as u64 == num_blocks,
+                || self.blocks.len() as u64 == retrieval_request.num_blocks(),
             "not enough blocks returned, expect {}, get {}",
-            num_blocks,
+            retrieval_request.num_blocks(),
             self.blocks.len(),
         );
+        ensure!(
+            self.status != BlockRetrievalStatus::SucceededWithTarget
+                || retrieval_request.match_target_id(self.blocks.last().unwrap().id()),
```

Comment:
we can't do unwrap here, otherwise a request with no blocks and SucceededWithTarget would panic the client.